lib/portage/eapi.py

```diff
@@ -324,113 +203,124 @@ def eapi_has_sysroot(eapi):
 )
 
 
-_eapi_attr_func_prefixes = (
-    "eapi_allows_",
-    "eapi_has_",
-    "eapi_requires_",
-    "eapi_supports_",
-    "eapi_",
-)
-
-
-def _eapi_func_decorator(func, attr_getter):
-    def wrapper(eapi):
-        return attr_getter(_get_eapi_attrs(eapi))
-
-    wrapper.func = func
-    wrapper.__doc__ = func.__doc__
-    return wrapper
-
-
-def _decorate_eapi_funcs():
-    """
-    Decorate eapi_* functions so that they use _get_eapi_attrs(eapi)
-    to cache results.
-    """
-    decorated = {}
-    for k, v in globals().items():
-        if not (
-            isinstance(v, types.FunctionType) and k.startswith(_eapi_attr_func_prefixes)
-        ):
-            continue
-        for prefix in _eapi_attr_func_prefixes:
-            if k.startswith(prefix):
-                attr_name = k[len(prefix) :]
-                if hasattr(_eapi_attrs, attr_name):
-                    decorated[k] = _eapi_func_decorator(
-                        v, operator.attrgetter(attr_name)
-                    )
-                    break
-    globals().update(decorated)
-
-
-_decorate_eapi_funcs()
-
-
-_eapi_attrs_cache = {}
-
-
-def _get_eapi_attrs(eapi):
+class Eapi:
+    ALL_EAPIS = (
+        "0",
+        "1",
+        "2",
+        "3",
+        "4",
+        "4-python",
+        "4-slot-abi",
+        "5",
+        "5-progress",
+        "6",
+        "7",
+        "8",
+    )
+    EAPI_RANGE_MAP = {
+        0: (0, 1),
+        1: (1, 2),
+        2: (2, 3),
+        3: (2, 4),
+        4: (4, 7),
+        5: (7, 9),
+        6: (9, 10),
+        7: (10, 11),
+        8: (11, 12),
+    }
+
+    _eapi_val: int = -1
+
+    def __init__(self, eapi_string: str):
+        if not eapi_string in self.ALL_EAPIS:
+            raise ValueError(f"'{eapi_string}' not recognized as a valid EAPI")
+        self._eapi_val = int(eapi_string[0])
```

Comment:
This `int(eapi_string[0])` usage seems fragile, since we will eventually have EAPI 10 with two digits instead of one. Maybe use `int(eapi_string.partition("-")[0])`.

Reply:
Great point, I'll make the suggested change.

Reply:
...actually, I think I'll go with a regex, i.e. `^(\d+)`, otherwise the singular strings, i.e. `"8"` would not match.

another option is to first try converting the string as-is, i.e. `try: int(eapi_string)`, but I'm leaning more towards the regex solution. LMK if you have an opinion.

Reply:
Note that `int(eapi_string.partition("-")[0])` actually does work with a singular string like `"8"`, but I suppose a regular expression is also fine.

Reply:
Ah, great point!

I think my unfamiliarity with python is showing 😅

I'll abandon the regex approach, it's a pretty big hammer to use when clearly it is not needed. thanks for the help!